### ISSUE TITLE
Apply function generics as part of type definition

### ISSUE THREAD

**Issue by jmacmahon** [microsoft/TypeScript#40542]
## Search Terms

generic typeof functions

## Suggestion

The problem is with the code as below:
```ts
const id = <T>(val: T) => val
type IdType = typeof id
// type IdType = <T>(val: T) => T

type Ret = ReturnType<typeof id>
// type Ret = unknown
```

It would be nice to be able to do something like so:
```ts
const id = <T>(val: T) => val
type IdType<T> = (typeof id)<T>
// type IdType<T> = (val: T) => T

// Or by extension:
type Ret<T> = ReturnType<(typeof id)<T>>
// type Ret<T> = T
```

## Use Cases

This would be useful to use with the [vriad/zod](https://github.com/vriad/zod) library, to implement different field validators depending on a parameter, like so:

```ts
// the dateSchema param could be a schema for an ISO 8601 string or a Date object, or even some custom date-like type
const fooSchema = <T>(dateSchema: z.ZodSchema<T>) => z.object({
  name: z.string(),
  date: dateSchema
})
type Foo<T> = z.infer<ReturnType<(typeof fooSchema)<T>>>
// expected result: type Foo<T> = { name: string, date: T }
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by Gozala** [comment]
I am understanding this proposal correctly that `IdType` here

```
const id = <T>(val: T) => val
type IdType = typeof id
```

would be `type IdType<T> = (val:T) => T` as opposed to `type IdType = <T>(val:T) => T` ? Or is idea here that you you should be able to qualify generic function type like `type IdType = <T>(val:T) => T` via `<T>`  suffix e.g. `type IdNum = IdType<number> // (val:number) => number` ?

**Comment by jmacmahon** [comment]
I meant the latter, so like you say, `type IdNum = IdType<number> // (val:number) => number`.

I think the former would be incorrect, because the parameter `T` is not present when declaring `type IdType`.

**Comment by ahejlsberg** [comment]
This feature is now implemented in #47607.
